openshift-knative/serverless-operator
# [wip] [SRVKS-900] Backport env override in upstream operator

## Proposed Changes



- Enables overriding of env vars in Knative operator.
- Backports https://github.com/knative/operator/pull/1085. The approach here is preferred compared to having an annotation on CR https://github.com/knative/operator/pull/1083, (note that the env setting cannot be replaced with a setting in config-net-istio as we need to apply informer filtering when net-istio controller starts at the sharedmain call).
- Patching the crds shouldnt create compatibility issues as the fields should be preserved if operator is downgraded. 
Also in the past the crds have been patched multiple times eg. Resource overrides.
- Discussion for backporting [here](https://coreos.slack.com/archives/CF5ANN61F/p1654114246116849) as it is hard to backport upstream anyway.
- This PR can be reverted once operator moves to 1.5 (we are a few versions behind).
- To test:
``` 
export KO_DOCKER_REPO=...
make images
make install-full-mesh
FULL_MESH=true INSTALL_SERVING="false" INSTALL_EVENTING="false" ./hack/install.sh
```
```
Apply:

apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  ingress:
    istio:
      enabled: true
  deployments:
    - annotations:
        sidecar.istio.io/inject: "true"
        sidecar.istio.io/rewriteAppHTTPProbers: "true"
      name: activator
    - annotations:
        sidecar.istio.io/inject: "true"
        sidecar.istio.io/rewriteAppHTTPProbers: "true"
      name: autoscaler
    - annotations:
        test: "true"
      name: net-istio-controller
    - env:
        - container: controller
          envVars:
            - name: ENABLE_SECRET_INFORMER_FILTERING_BY_CERT_UID
              value: "true"
      name: net-i
```
Observe deployment env:

```
$ oc get pods  net-istio-controller-57dff5786d-sq79s     -n knative-serving -o jsonpath='{.spec.containers[].env}' | jq .
[
  {
    "name": "SYSTEM_NAMESPACE",
    "valueFrom": {
      "fieldRef": {
        "apiVersion": "v1",
        "fieldPath": "metadata.namespace"
      }
    }
  },
  {
    "name": "CONFIG_LOGGING_NAME",
    "value": "config-logging"
  },
  {
    "name": "CONFIG_OBSERVABILITY_NAME",
    "value": "config-observability"
  },
  {
    "name": "METRICS_DOMAIN",
    "value": "knative.dev/net-istio"
  },
  {
    "name": "ENABLE_SECRET_INFORMER_FILTERING_BY_CERT_UID",
    "value": "true"
  }
]
```
Full test using the net-istio issue [here](https://gist.github.com/skonto/6010c0da420b044dd768927408f9472b).
